### ISSUE TITLE
Use num2words for universal numbers to words converter in Festival

### ISSUE THREAD

**Issue by Pallas1303** [festvox/festvox#36]
I can use [num2words](https://pypi.org/project/num2words/) for automatic numbers words converter. Using (system) function, in my function **(define (festpb_pt::number token name)**

```scheme
(define (festpb_pt::number token name)
  "(festpb_pt::number token name)
Return list of words that pronounce this number in pt."

;Set name random, using rand function
(set! ra (rand))

;Set command for system function, piper output of num2words in file with name of variable ra 
(set! command (format nil "num2words -l pt-BR %s > %s" name ra))

;Execute command using system function
(system command)

;Load file how a list with load function
(load ra t)
)
```

For deletion of cache file have function in [Festvox Doc](http://festvox.org/docs/manual-2.4.0/festival_34.html#:~:text=(delete%2Dfile%20FILENAME)) with this in my function **(festpb_pt::number token name)**

```scheme
(delete-file ra)
```

But my function, not return when have this function.

**Comment by ddavout** [comment]
a function with no return: that remembers me the bash functions 
It sounds (to me..) that you need the help of a file
but even using `festival-freebsoft-utils`
I have nothing simpler than

```
      ;; debug
      (set_backtrace t)
      (require 'fileio)
```

```
(define (fest_pt::number number lang to)
  "Return list of words that pronounce this number in lang pt-PT.
  to cardinal currency ordinal"

    (let (command tfn result)

        ; prepare a temporary file
        (set! tfn (make_tmp_filename));
        (set! command (format nil "num2words -l %s --to %s %s >%s" lang to number tfn))
        ;; debug ;  Print DATA to stdout if textual form.  Not a pretty printer
        (print command)
        ; issue the command
        (system command)
            ;; not successful:   would only the first word
            ; ;Read and return next s-expression from file indicated by the file pointer createf by fopen
            ;; the file pointer not the file itself
            ;; otherwise criptic error message SIOD ERROR: not a file : "./tmpdir/est_2836297_00005"
         ; (set! fdp (fopen tfn "r"))
         ;(set! resultat  (readfp fd))
        ; (fclose fdp)
          
        ; using read-file de 'fileio
        (set! result (read-file tfn))
        ; dispose
        ;; (delete-file tfn)
        result))
```

(set! number (* 1000 (rand)))
> 394.38293
(fest_pt::number number  "pt_PT" "currency") will give
"trezentos e noventa e quatro euros e trinta e oito cêntimos
"
with "cardinal" 
"trezentos e noventa e quatro vírgula três oito dois nove três

**Comment by Pallas1303** [comment]
Thanks for your helping. But I not understand your code, why with coins? When using the (delete-file) function in the function I mentioned. do not generate the pronunciation. the expected is to return a list of words, with this I used the function (load t). to spoken to be used to synthesize.

**Comment by ddavout** [comment]
Sorry, for the misunderstanding, I should have renamed the function !!!
The aim of mine is to not generate the pronunciation but to generate the string to pronounce.
and it would be used in yours. ( that is , I must say, rather confusing ..)

I did use the option --to of num2words to test it because it's of interest when we want to read currencies
( by the way,  I've got 
`festival> (fest_pt::number "10"  "fr_FR" "currency");`
"dix euros et zéro centimes
as withe  pt_PT  you have
"dez euros )

**Comment by Pallas1303** [comment]
I understand, I'm doing a voice for my language. Have various files in festvox folder with settings and features of my voice (how: phoneset,  dic). 

This case the function is execution correct on terminal of Festival. But in my voice in festpb_pt_pieroni_tokenizer.scm or for major understanding (INST_LANG_VOX_tokenizer.scm).

The function (define (INST_LANG::number token name) line 60 not return pronunciation of numbers when used (delete-file) function.

**Comment by ddavout** [comment]
but with 
`(set! command (format nil "num2words -l pt-BR %s > %s" "4"  ra))
"num2words -l pt-BR 4 > 0.78309922"
you don''t create a file`

**Comment by Pallas1303** [comment]
>
> ```scheme
> ;Set name random, using rand function
> (set! ra (rand))
> 
> ;Set command for system function, piper output of num2words in file with name of variable ra 
> (set! command (format nil "num2words -l pt-BR %s > %s" name ra))
> 
> ;Execute command using system function
> (system command)
> 
> ;Load file how a list with load function
> (load ra t)
> )
> ```

**Comment by ddavout** [comment]
ok, it's the use of a flonum as a filename that confused me
I was lost in my own problem with this delete-file
( it's used in vanilla festvox in build_clustergen.scm, 
if the work is done in parallel)
Maybe the cache is not cleared at the right time ;)

**Comment by Pallas1303** [comment]
I am building a clunits voice (Unit Selection) with corpus of phrases phonetically balanced. With clustering voice have some problems for building. 

This problem for me is, a that of function output is a list de words when is used (delete-file) the function (this case is numbers to words)  return nothing . I not know how fix this "problem".

**Comment by ddavout** [comment]
I try your code on a artic voice, 
I checked thge permissions on ra with
(set! command_l (format nil "/usr/bin/ls -al %s" ra ))
but what I don't understand is  where exactly you can put this delete-file   **inside** your function ? , at best you don't send anything or raise a SIOD error, no ?
if instead of ra, you use a prefixed file,  you could clean the cache beforehand  (?)

**Comment by Pallas1303** [comment]
About ra, this is my method for generation of a cache file. Because, I didn't see the documentation for the function (make_tmp_file) but is just redo the how variable for cache name file will be. Your tried in a English Voice or in other language? For English, the Festival maked this convention.

Strangely, without SIOD errors.

**Comment by ddavout** [comment]
I tried with a vanilla cmu_us_ksp_arctic
What I know that delete-file doesn't give a SIOD error in (define (parallelSystemFlush) in clustergen_build.scm
but a stratagem is used to postpone the system calls,

Envoyé avec la messagerie sécurisée [Proton Mail.](https://proton.me/)

**Comment by Pallas1303** [comment]
I understand, this is some thing of Festival. I will repost this for repository of Festival Speech Synthesis.

**Comment by ddavout** [comment]
I don't want to pollute your new thread, but I would like to try  again to be clear, give a chance


`(set! tmp)

(define (essai number tmp)
 (set! tmp (make_tmp_filename))
 (set! command (format nil "echo %s > %s" number tmp))
 (system command)
 (load tmp t)
 )
 
 (essai 3 tmp)
 (if (probe_file tmp) (delete-file tmp))
 (essai 4 tmp)
(delete-file tmp)`

would work, but if you put  a delete-file or nearly anything else inside the essai function after the (load tmp t)
the result of your function would not be (load tmp t)

**Comment by Pallas1303** [comment]
Tranks for your help, but I not have how your code. had a problem in my ambient of development and to solve this problem I will take quite a long time.

I am thinking, defined a tmp before of execute function. The function, with more a argument.

**Comment by ddavout** [comment]
I've just use echo instead of num2words to make it easier to others
I made tmp a global variable,  you can choose to make a function whatever but it has to be outside the function as delete-file spoils the result (sending something else or raising a SIOD as you can't load a deleted file)
